### ISSUE TITLE
EditorConfig UI should not use `value:severity` syntax

### ISSUE THREAD

**Issue by Youssef1313** [dotnet/roslyn#52694]
See the following test for example:

https://github.com/dotnet/roslyn/blob/b21007bf0b5206eaee41245a0a9f28684254bbb6/src/EditorFeatures/CSharpTest/EditorConfigSettings/Updater/SettingsUpdaterTests.cs#L72-L79

Related to: https://github.com/dotnet/roslyn/issues/44201

cc @jmarolf

**Comment by jmarolf** [comment]
Need to discuss a design for this. As far as I know we haven't committed to deprecating this syntax

**Comment by Youssef1313** [comment]
This syntax is recognized in IDE only and doesn't work for command line builds.

**Comment by jmarolf** [comment]
today while this form
```ini
[*.cs]
dotnet_style_qualification_for_field = true:warning
```

can be tranlated to this:
```ini
[*.cs]
dotnet_diagnostic.IDE0009.severity = warning
```

there is nothing that can be done for this form
```ini
[*.cs]
dotnet_style_qualification_for_field = false:warning
```

once https://github.com/dotnet/roslyn/issues/44201 is closed (meaning that all codestlye analysis accepts some custom option such as `dotnet_diagnostic.IDE0009.enabled = false`) the UI can be updated to write elements out in this style

**Comment by CyrusNajmabadi** [comment]
Can you include me in any convos here.  For one thing, i'm very wary about the fact that this approach seems to encode IDE diag ids as first class citizens.  That means we're beholden to preserving those exact IDs with those exact semantics.  That's not something i would have wanted as we very much would like to be able to change internal impl details on how an analyzer/fixer communicates, without that being something that is part of a public contract.  the benefit of things like `dotnet_style_qualification_for_field` (beyond just being human readable) is that is relaxes the internal impl for anlayzer/fixer.  We could have a single human-facing option map to many IDs under the cover to support whatever approach the analyzer/fixer wanted to take.

**Comment by jmarolf** [comment]
@CyrusNajmabadi this was discussed here: https://github.com/dotnet/roslyn/issues/44201 though I don't think either of us were at that meeting. We should bring this back up in the design meeting. I also have concerns here

**Comment by mavasani** [comment]
> Need to discuss a design for this. As far as I know we haven't committed to deprecating this syntax

THat is not correct. https://github.com/dotnet/roslyn/issues/44201 was brought up in the design meeting in the past, and we decided to go ahead with the deprecation. The conclusion was that once with have editorconfig UX, it is fine to have separate entries and syntax for configuring option value and severity.

Additionally, the compiler team has explicitly called out that only way that they will support severity configuration of any analyzer diagnostics is via the `dotnet_diagnostic` syntax, so unless we are able to convince the compiler team here to respect IDE specific severity syntax, that approach is a non-starter.

**Comment by mavasani** [comment]
> i'm very wary about the fact that this approach seems to encode IDE diag ids as first class citizens

Note that using the `option_value:severity` is actually much worse in many more aspects:
1. Forces compiler to now understand this additional syntax for severity configuration
2. Compiler needs to somehow figure out how each IDE code style option name maps to IDExxxx diagnostic ID, otherwise there is no way for it to apply the severity to reported code style diagnostics.
3. Syntax for configuring IDE code style diagnostic severity will diverge from those for every other analyzer out there, including the first party .NET CA analyzers. That is extremely unfortunate.

Overall, I think there are multiple reasons why we should not be trying to special case IDE code style analyzers here - the recommendation and rule has always been very straightfoward for all analyzers:
1. Use `option_name = option_value` entries in editorconfig for your option configuration
2. Use `dotnet_diagnostic.ID.severity = severity` for configuring severity, which is specifically a diagnostic specific concept.
3. In past, before editorconfig was supported in compiler, analyzer severity can be adjusted via `ruleset` files for configuring analyzer severity for diagnostic IDs. `/nowarn` and `/warnaserror` have special configuration semantics.

`:severity` syntax has always been a non-recommended approach from core analyzer/compiler viewpoint.

**Comment by mavasani** [comment]
> i'm very wary about the fact that this approach seems to encode IDE diag ids as first class citizens. That's not something i would have wanted as we very much would like to be able to change internal impl details on how an analyzer/fixer communicates

That is actually not true, in fact the reverse holds - analyzer options are opaque internal details for compiler, that only each analyzer should know. Diagnostic IDs are the only first class citizens that the compiler/core analyzer driver understands about diagnostics, and each analyzer needs to support these as first class citizens, they cannot pretend for it to be internal implementation details that can be changed.

**Comment by mavasani** [comment]
> the benefit of things like dotnet_style_qualification_for_field (beyond just being human readable) is that is relaxes the internal impl for anlayzer/fixer. We could have a single human-facing option map to many IDs under the cover to support whatever approach the analyzer/fixer wanted to take.

That cannot work - unless you hardcode such a map into the compiler layer. Additionally, it gives special status to IDE code style analyzers in that they are now allowed to have special syntax for severity configuration, that no other analyzer can use for their options. I don't believe such an approach is maintainable.

**Comment by mavasani** [comment]
Also tagging @sharwell - we have discussed this issue multiple times in the past.

**Comment by mavasani** [comment]
@CyrusNajmabadi Adding one more point that the IDExxxx diagnostic IDs are not implementation details. These are officially documented and maintained in a tabular fashion at docs.microsoft.com: https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/#index

**Comment by mavasani** [comment]
> his was discussed here: #44201 though I don't think either of us were at that meeting.

@jmarolf I believe both @CyrusNajmabadi and @sharwell were at the design meeting where I brought this up and the deprecation was approved as the only possible path forward.

**Comment by sharwell** [comment]
The tl;dr for what @mavasani describes is the `value:severity` will only work if the compiler changes their **.editorconfig** implementation to recognize the special IDE configuration strings.

If we do this, the IDE analyzers will be configured in one way while every other analyzer is configured a different way. Since the compiler isn't planning on making this change, the `value:severity` is effectively deprecated by the simple fact that it doesn't currently work and there are no plans to make it start working.

**Comment by CyrusNajmabadi** [comment]
> Compiler needs to somehow figure out how each IDE code style option name maps to IDExxxx diagnostic ID, otherwise there is no way for it to apply the severity to reported code style diagnostics.

Doesn't that mapping have to exist *somewhere*?  After all, how do we expect the UI to work here?  How does it discover this mapping?

> The tl;dr for what @mavasani describes is the value:severity will only work if the compiler changes their .editorconfig implementation to recognize the special IDE configuration strings.

I think that makes a ton of sense.  this approach feels much more pleasant for users as there's just a single, easy to understand, name they can put in their editor config files.  The important parts of this can then be set by the user right there.  There's a clear connection between this entry and any sort of UI, instead of having the UI now affecting multiple disparate strings in the .editorconfig that seem unrelated.

Is there a reason compiler would not want to do this?

> while every other analyzer is configured a different way. 

I would presume if compiler supported this, other analyzers would be able ot benefit from this as well.

**Comment by CyrusNajmabadi** [comment]
>  they cannot pretend for it to be internal implementation details that can be changed in future analyzer releases.

This was definitely not understood well at the time these features went into implementation.  I get that this is something we're now seemingly trying to be firmer on, but it honestly seems like a less pleasant experience for users (albeit with engineering simplicity on our side).

Trust me, i totally get why using the IDE just removes a whole host of questions.  it just seems highly suboptimal to me from the perspective of ever having anyone interact with these editorconfig files.  If i think of the file as just omething you never touh and only use a UI for.  Fine, then it's ok.  but at that point, we could just make huge blobs of xml that we coudl encode things with.  My presumption is that we should design for the nicest experience for the user when both hand-authoring *and* reading those files.  And once we hve that design, we should push through whatever engineering changes are needed to support that.

Thanks! :)

**Comment by mavasani** [comment]
@sharwell Thanks for nicely summarizing my points

@CyrusNajmabadi Essentially it boils down to this: We need a compiler feature to support another syntax for specifying diagnostic severities in editorconfig. Given that the syntax/entry that we want to push for does not have any specification of the diagnostic ID or category, but just an option name, we need an additional compiler feature request to allow analyzers to provide a mapping from its supported options to diagnostic descriptors, so that the compiler can decode the `option_name = option_value:severity` entries, and map the `option_name` to the diagnostic ID whose severity needs to be configured.

As @sharwell clarified, the above feature requested was rejected by the compiler team on the ground that there already exists a supported syntax for specifying severities `dotnet_diagnostic.ID.severity = %severity%`. Additionally, given our investment into an editorconfig UX that should do all the reading/writing of severities for the user, these feature requests seem redundant and push users towards manually editing severities in editorconfig as opposed to let the editorconfig UX do the right thing. However, if you can still convince the compiler team towards implementing this feature, I have no strong objection.

**Comment by mavasani** [comment]
> Doesn't that mapping have to exist somewhere? After all, how do we expect the UI to work here? How does it discover this mapping?

Yes, but this is currently internal to the IDE layer: https://github.com/dotnet/roslyn/blob/main/src/Analyzers/Core/Analyzers/IDEDiagnosticIdToOptionMappingHelper.cs. As soon as we want to push it down to the compiler, this would need a first class feature support for specifying mapping between diagnostic ID/descriptor to supported option names.

**Comment by jmarolf** [comment]
@mavasani I am sorry if I missed it but what is the equivalent to this?

```ini
[*.cs]
dotnet_style_qualification_for_field = false:warning
```

I do not see this mentioned anywhere in the dicussions.

**Comment by sharwell** [comment]
@jmarolf @Youssef1313 More specifically, only IDE0003 is used when the value is false, and only IDE0009 is used when the value is true. You don't need to configure the severity of both IDE0003 and IDE0009 when the value of `dotnet_style_qualification_for_field` is configured in the same **.editorconfig** section.

**Comment by Youssef1313** [comment]
@sharwell The EditorConfig UI will have no way telling which value used by which diagnostic id, so it will have to set the severity for both.

**Comment by jmarolf** [comment]
sorry for being abtuse, so the compiler understands
```ini
[*.cs]
dotnet_style_qualification_for_field = false
```

and this will work on commandline builds but we can't get
```ini
[*.cs]
dotnet_style_qualification_for_field = false:warning
```

to work?

**Comment by CyrusNajmabadi** [comment]
```
dotnet_style_qualification_for_field = false
dotnet_diagnostic.IDE0003.severity = warning
dotnet_diagnostic.IDE0009.severity = warning
```

Can we make sure that if we emit that, we include something in the line to make it clear what the two severity lines correspond to?

**Comment by sharwell** [comment]
> The EditorConfig UI will have no way telling which value used by which diagnostic id, so it will have to set the severity for both.

The UI understands anything we code it to understand (and in this case we own both the UI and the semantics for all impacted settings).



> sorry for being abtuse, so the compiler understands
> 
> ```ini
> [*.cs]
> dotnet_style_qualification_for_field = false
> ```
> 
> and this will work on commandline builds but we can't get
> 
> ```ini
> [*.cs]
> dotnet_style_qualification_for_field = false:warning
> ```
> 
> to work?

Technically the compiler doesn't understand either of these, and under the current design doesn't need to. Starting with #43546, the compiler doesn't run analyzers that are marked as `hidden` or `info`. If the analyzer has one of these as the defaults and the configuration doesn't raise the severity via a compiler-recognized `dotnet_diagnostic` line, the analyzer will not execute at all and therefore will not report any diagnostics at any severity.

**Comment by jmarolf** [comment]
> Technically the compiler doesn't understand either of these, and under the current design doesn't need to. 

To connect the dot for me, the compiler doesn't need to know about `dotnet_style_qualification_for_field`  because analyzers in the code style layer read this option directly?

**Comment by sharwell** [comment]
> > Technically the compiler doesn't understand either of these, and under the current design doesn't need to.
> 
> To connect the dot for me, the compiler doesn't need to know about `dotnet_style_qualification_for_field` because analyzers in the code style layer read this option directly?

Correct. And the analyzers only run if the compiler has been instructed to run them via a severity configuration of `warning` or `error`.

**Comment by jmarolf** [comment]
alright, I am now fine updating the UI to do this :)

**Comment by jmarolf** [comment]
Had I been in the meeting 1+ years ago I would have pointed out that we could just add a an API where the compiler calls the analyzer and gives it the dictionary with the applicable editorconfig string and the analyzer can just return true/false for whether it should be run. Then parsing this option type doesn't need to live in the compiler codebase and we don't need to change every editorconfig file on earth.

**Comment by mavasani** [comment]
@jmarolf @sharwell Do we want to bring this up at today's design meeting? I can give a quick overview/history of how this issue has progressed over the lifetime of "editorconfig support in compiler" feature work and we can confirm if we want to go ahead with deprecating the `:severity` suffix OR go the route of compiler feature requests for it to understand it.

**Comment by jmarolf** [comment]
yeah, my remaining question is how we see naming styles being enforced on build in the future? I assume that was also discussed originally.

**Comment by mavasani** [comment]
> my remaining question is how we see naming styles being enforced on build in the future

Yes, this has also been discussed in the past. Naming styles can already be enforced on build today, see note on the severity section in https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/naming-rules#naming-rule-properties. Currently, all the naming rules share a single ID, so you cannot have different severities for different naming styles. There have been 2 alternative solutions proposed for it:
1. Use different diagnostic IDs for different buckets/each naming rule
2. Sam's proposal for sub-diagnostic IDs: https://github.com/dotnet/roslyn/issues/27712

I personally like Sam's proposal as it solves multiple issues/feature requests at once, but it is more work, including new feature in the compiler layer.

**Comment by jmarolf** [comment]
I think we can have a quick discussion to make sure everyone is still happy with the direction. but this answers my questions.

**Comment by CyrusNajmabadi** [comment]
I'm happy if we can provide a good human experience, specifically around understanding and maintaining the .editorconfig file.  If that's doable, (even it's just through things like comments) then i'm ok with a deprecation plan.

**Comment by CyrusNajmabadi** [comment]
Also, if we do deprecate thsi:

1. our tooling shoudl let users know
2. our tooling should migrate the settings to the new form for the user.

**Comment by jasonmalinowski** [comment]
Per the design meeting discussion, we agreed we should make the tool do what it has to do for now, but we really should revisit the discussion to deprecate the syntax for specifying severities. This particular motivating example:

```
dotnet_style_qualification_for_field = false
dotnet_diagnostic.IDE0003.severity = warning
dotnet_diagnostic.IDE0009.severity = warning
```

Might be _correct_ but it was hard to explain why this was really a good experience, since we would indeed need to add comments to the file to explain to users what is going on here. @mavasani had an idea of how we can make this work, and offered to write up a proposal.

**Comment by sharwell** [comment]
@mavasani can you update this to reflect the new design?

**Comment by mavasani** [comment]
Sure, I think EditorConfig UX is already respecting `:severity` syntax and also using it for writing out changes made in the UX into the editorconfig file. I'll take a closer tab at current implementation and behavior and confirm if we need to make any changes here.
